proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+## Effect and Interactions
+
+As discussed several times throughout this proposal, its goal is to give a clear and simple specification of impredicative instantiation within GHC. Whereas "simple" is a subjective matter, the availability of a set of typing rules defines a clear specification.
+
+### `TypeApplications`
+
+We deem the interaction with `TypeApplications` as a very important one; the goal is for our specification to benefit from user-written types as much as possible. We want `(:) @(forall a. a -> a) (\x -> x) []` to work, without the need of an additional type application in `[]`. The draft paper details the interaction between those features in depth.
+
+### `RankNTypes` and lack of contravariance
+
+The main disadvantage of making "quick look impredicativity" official as described in the [paper draft](https://www.dropbox.com/s/hxjp28ym3lptmxw/quick-look-steps.pdf?dl=0) is that it collides with *contravariance of function types*. That is, it no longer holds that:
+
+`(Int -> Int) -> R is a subtype of (forall a. a -> a) -> R`
+
+The problem is of technical nature, but without this restriction we would not be able to infer impredicative instantiations of functions like `($)`. **This restriction affects not only users of `ImpredicativeTypes`, but also users of `RankNTypes`, effectively making this proposal not 100% backwards compatible.**
+
+Note anyway that the support for contravariance in GHC is really fragile. Take for example the functions `choose :: forall a. (a -> Bool) -> (a -> Bool) -> Bool`, `f :: (Bool -> Bool) -> Bool` and `g :: (forall a. a -> a) -> Bool`. In theory we can accept the expressions `choose f g` and `choose g f` if we instantiate `a` with `(forall a. a -> a)`. Here is an excerpt of a GHCi session which shows that the one of the expressions is accepted and the other is not:
+
+```text
+Prelude> :set -XRankNTypes -XImpredicativeTypes
+Prelude> choose :: (a -> Bool) -> (a -> Bool) -> Bool ; choose = undefined
+Prelude> f :: (Bool -> Bool) -> Bool ; f = undefined
+Prelude> g :: (forall a. a -> a) -> Bool ; g = undefined
+Prelude> :t choose f g
+
+<interactive>:1:10: error:
+    • Couldn't match type ‘Bool -> Bool’ with ‘forall a. a -> a’
+      Expected type: (Bool -> Bool) -> Bool
+        Actual type: (forall a. a -> a) -> Bool
+    • In the second argument of ‘choose’, namely ‘g’
+      In the expression: choose f g
+Prelude> :t choose g f
+choose g f :: Bool
+```
+
+There is a simple workaround that works most of the time: eta-expansion (you might need to eta-expand more than once, though). Taking the `choose`, `f`, and `g` functions defined above,in a non-contravariant world we need to eta-expand `f` to make it work: `choose (\x -> f x) g`; now the `x` is assigned type `forall a. a -> a`, which is then instantiated to `Int -> Int` before passing it to `f`.
```

Comment:
Is eta-expansion a workaround under the current sytem, under this proposal, or both?